render/host.go

```diff
@@ -62,18 +59,73 @@ func MapX2Host(n report.Node, _ report.Networks) report.Nodes {
 	return result
 }
 
-// MapEndpoint2Host takes nodes from the endpoint topology and produces
+// endpoints2Hosts takes nodes from the endpoint topology and produces
 // host nodes or pseudo nodes.
-func MapEndpoint2Host(n report.Node, local report.Networks) report.Nodes {
-	// Nodes without a hostid are treated as pseudo nodes
-	hostNodeID, timestamp, ok := n.Latest.LookupEntry(report.HostNodeID)
-	if !ok {
-		return MapEndpoint2Pseudo(n, local)
+type endpoints2Hosts struct {
+}
+
+func (e endpoints2Hosts) Render(rpt report.Report, dct Decorator) report.Nodes {
+	ns := SelectEndpoint.Render(rpt, dct)
+	local := LocalNetworks(rpt)
+
+	var ret = make(report.Nodes)
+	var mapped = map[string]string{} // input node ID -> output node ID
+	for _, n := range ns {
+		// Nodes without a hostid are treated as pseudo nodes
+		hostNodeID, timestamp, ok := n.Latest.LookupEntry(report.HostNodeID)
+		if !ok {
+			id, ok := pseudoNodeID(n, local)
+			if !ok {
+				continue
+			}
+			addToResults(n, id, ret, mapped, func() report.Node {
+				return report.MakeNode(id).WithTopology(Pseudo)
+			})
+		} else {
+			id := report.MakeHostNodeID(report.ExtractHostID(n))
+			addToResults(n, id, ret, mapped, func() report.Node {
+				return report.MakeNode(id).WithTopology(report.Host).
+					WithLatest(report.HostNodeID, timestamp, hostNodeID)
+			})
+		}
 	}
+	fixupAdjancencies(ns, ret, mapped)
+	return ret
+}
+
+// Add Node M to the result set ret under id, creating a new result
+// node if not already there, and updating the old-id to new-id mapping
+// Note we do not update any counters for child topologies here
```

Comment:
That last bit explains what it does but should really explain why.

Is it because of performance? Is that actually noticeable? And why is it safe? Because in current usage `m` is always an endpoint, and we don't depend on endpoint counters anywhere?

Reply:
It added 4-5% to the number of allocations, and makes no difference to the output.
I did not get the bottom of why it makes no difference.
`m` is a `process` in one case; `endpoint` in two cases.

Reply:
`m` is always an endpoint.

Reply:
which is why it makes no difference.

Reply:
Sorry, you're right, it's always an endpoint